lib/Dialect/SV/SVOps.cpp

```diff
@@ -817,23 +817,52 @@ void WireOp::getAsmResultNames(OpAsmSetValueNameFn setNameFn) {
 }
 
 // If this wire is only written to, delete the wire and all writers.
-LogicalResult WireOp::canonicalize(WireOp op, PatternRewriter &rewriter) {
-  // If the wire is named, then we can't delete it.
-  if (!op.name().empty())
+LogicalResult WireOp::canonicalize(WireOp wire, PatternRewriter &rewriter) {
+  // If the wire has a symbol, then we can't delete it.
+  if (wire.sym_nameAttr())
     return failure();
 
-  // Check that all operations on the wire are sv.connects. All other wire
-  // operations will have been handled by other canonicalization.
-  for (auto &use : op.getResult().getUses())
-    if (!isa<ConnectOp>(use.getOwner()))
+  // Wires have inout type, so they'll have connects and read_inout operations
+  // that work on them.  If anything unexpected is found then leave it alone.
+  SmallVector<sv::ReadInOutOp> reads;
+  sv::ConnectOp write;
+
+  for (auto *user : wire->getUsers()) {
+    if (auto read = dyn_cast<sv::ReadInOutOp>(user)) {
+      reads.push_back(read);
+      continue;
+    }
+
+    // Otherwise must be a connect, and we must not have seen a write yet.
+    auto connect = dyn_cast<sv::ConnectOp>(user);
+    // Either the wire has more than one write or another kind of Op (other than
+    // ConectOp and ReadInOutOp), then can't optimize.
+    if (!connect || write)
       return failure();
+    write = connect;
+  }
 
-  // Remove all uses of the wire.
-  for (auto &use : make_early_inc_range(op.getResult().getUses()))
-    rewriter.eraseOp(use.getOwner());
+  Value connected;
+  if (!write) {
+    // If no write and only reads, then replace with XOp.
+    connected = rewriter.create<ConstantXOp>(
+        wire.getLoc(),
+        wire.getResult().getType().cast<InOutType>().getElementType());
+  } else if (isa<hw::HWModuleOp>(write->getParentOp()))
+    connected = write.src();
```

Comment:
I'd recommend making this a little bit stronger: you can eliminate the store if the **stored value** is global, something like:

```
 } else {
   connected = wire.src();
   if (auto srcOp = connected.getDefiningOp())
     if (!isa<hw::HWModuleOp>(srcOp->getParentOp)) 
       return failure();
}
```

should do it.  This will allow ports and values at the top level to be optimized

Reply:
This begs the question of what the semantics this is:
```
wire a
b = exp(...)
sv.if() {
  connect a, b;
}
```

Reply:
X until proven otherwise.

Reply:
Z would seem more natural when the if isn't taken.

Reply:
Yeah, great point. My understanding is that assignments in an sv.if statement in an always block are equivalent to a mux, so it is equivalent to:

`connect a, (mux(cond, b, Z))`

in the always block.

Reply:
If the write is in an always block, Is this legal ?

```mlir
  hw.module @always_wire_canon1(%arg0: i1) -> (i1) {
    %true = hw.constant true
    %0 = sv.wire  : !hw.inout<i1>
    sv.always posedge %arg0  {
      sv.connect %0, %true : i1
    }
    %1 = sv.read_inout %0 : !hw.inout<i1>
    hw.output %1 : i1
  }
```
```mlir 
  hw.module @always_wire_canon1(%arg0: i1, %arg1: i1) -> (i1) {
    %true = hw.constant true
    hw.output %true : i1
  }
```

Reply:
Please commit with the more conservative check.
That said, the sections of the 2017 SV standard you want to look at are 4.9.1-4.9.4.  My understanding is you can have assign statements to wires inside procedural blocks, but this is deprecated and likely going away in the future (annex C).

sv.connect seems a vestage of firrtl and should probably be sv.assign to make it clear what it represents.

So, per SV, = and <= (passign and bpassign) operate only on reg type and not wire, assign operates on wire, and is legal inside procedural regions, but not recommended.  per 10.6.1, assign in a procedural region takes precedence over other assignments (in other procedural regions, see the example in 10.6.1).  We are not verifying wire v.s. reg operations right now (netlist vs variable), but should.

Ugg, 10.3.2 example 3 shows some crazy syntax (assign isn't a statement, but a block!).

Reply:
> sv.connect seems a vestage of firrtl and should probably be sv.assign to make it clear what it represents.

I agree we need to look at sv.connect again.  We need something that can be in graph regions (e.g. top level, sv.always_comb), using passign/bpassign in procedural regions is logical.  Renaming it could also make sense.

> Please commit with the more conservative check.

What is the concern here?  If you have one write to a wire, and if the written value is usable at the module level, then it should always be safe to substitute it, no matter if it is a bpassign, a passign, or a connect.  The only other path would have an x/z, right?

I suppose the problem there is that it would change the results of xprop/zprop, and a core transformation like canonicalize shouldn't do that.

Reply:
sv.always_comb is not a graph region, it is procedural.  One of the problems here is we treat wire and reg similarly, when they have fairly different rules in verilog (and we don't even have a way to distinguish the two at the port level).

If there is only one assign of a wire, but it is in a procedural region, it may have ordering constraints with reads in that region.  If the region it is in is a conditional, then we are making a wire that is sometimes driven always driven, which is a visible change and will break code (consider the wire drives an output from a module, you are introducing multi-driven wires in the parent module were before there were none).

Reply:
In the SV standard, although they have the same syntax, assign statements in a procedural region are a different op than assign statements at the module level.  (this is in addition to passign and bpassign).

Reply:
> sv.always_comb is not a graph region, it is procedural. 

Oh right, good point!

> we treat wire and reg similarly, when they have fairly different rules in verilog

Right.  They are syntactically similar, but actually completely different things.  wire is in the "logic" family.

I didn't realize you could have reg ports.  

> If there is only one assign of a wire, but it is in a procedural region, it may have ordering constraints with reads in that region.

Ok, but the reads would produce an x/z.  This is what I meant by it affecting xprop.

> consider the wire drives an output from a module, you are introducing multi-driven wires in the parent module were before there were none

Oh yes, excellent point, you're absolutely right.  This makes "z elimination" an invalid transformation, it isn't a UB like thing in C.  I'll need to reprogram some old habits.